File: RoboCartUtils/OccupancyGrid.cpp
#include "OccupancyGrid.hpp"

/*
 OccupancyGrid::OccupancyGrid(cv::Mat q, cv::Size occupancySize,
 double cameraHeight, double robotWidth, double robotLength, double clearance,
 cv::Range xRange, cv::Range yRange, cv::Range zRange,
 double r, double c, double deltaN, double nt, double lt, double deltaH, double wN, double wH);
 */

OccupancyGrid::OccupancyGrid() {
	cv::FileStorage fs("params.yml", cv::FileStorage::READ);
	cv::FileStorage ext("extrinsics.yml", cv::FileStorage::READ);
	cv::FileNode og = fs["og"];

	og["occupancySize"] >> occupancySize;
	cameraHeight = (double) og["cameraHeight"];
	robotWidth = (double) og["robotWidth"];
	robotLength = (double) og["robotLength"];
	clearance = (double) og["clearance"];
	og["xRange"] >> xRange;
	og["yRange"] >> yRange;
	og["zRange"] >> zRange;
	ext["Q"] >> q;
	r = (double) og["r"];
	c = (double) og["c"];
	deltaN = (double) og["deltaN"];
	nt = (double) og["nt"];
	lt = (double) og["lt"];
	deltaH = (double) og["deltaH"];
	wN = (double) wN;
	wH = (double) wH;
	count = 0;
}

void OccupancyGrid::compute(cv::Mat &disparity, cv::Mat &dispOccupancy,
		cv::Mat &image3d) {
	cv::reprojectImageTo3D(disparity, image3d, q, true);
	cv::Mat occupancy(occupancySize.height, occupancySize.width, CV_32FC1);
	occupancy = cv::Scalar(0);
	cv::Mat height(occupancySize.height, occupancySize.width, CV_32FC1);
	height = cv::Scalar(0);

	for (int i = 0; i < image3d.rows; i++) {
		for (int j = 0; j < image3d.cols; j++) {
			cv::Point3f pt = image3d.at<cv::Point3f>(i, j);
			double h = cameraHeight - pt.y;
			if (pt.x > xRange.end || pt.x < xRange.start || h > yRange.end
					|| h < yRange.start || pt.z > zRange.end
					|| pt.z < zRange.start) {
				continue;
			}

			float scaledZ = (pt.z - zRange.start) / (zRange.end - zRange.start);
			float scaledX = (pt.x - xRange.start) / (xRange.end - xRange.start);

			int row = occupancySize.height
					- floor(scaledZ * occupancySize.height);
			int col = floor(scaledX * occupancySize.width);
			occupancy.at<float>(row, col) += 1;
			height.at<float>(row, col) += h;
		}
	}

	cv::Mat test;
	normalize(occupancy, test, 0, 255, CV_MINMAX, CV_8UC1);

	dispOccupancy.create(occupancy.rows, occupancy.cols, CV_8UC3);
	dispOccupancy = cv::Scalar(0, 0, 0); //TODO: remove this, not needed.
	float xCam = occupancySize.width / 2, yCam = occupancySize.height;

	for (int i = 0; i < occupancy.rows; i++) {
		for (int j = 0; j < occupancy.cols; j++) {
			if (occupancy.at<float>(i, j) > 0) {
				float xPt = j, yPt = i;
				float distanceToCamera = sqrt(
						(xCam - xPt) * (xCam - xPt)
								+ (yCam - yPt) * (yCam - yPt));
				//cout << distanceToCamera << endl;
				float adjustedNum = occupancy.at<float>(i, j) * r
						/ (1 + exp(-distanceToCamera * c));
				float pijNum = 1 - exp(-(adjustedNum / deltaN));
				float lijNum = pijNum != 1 ? log(pijNum / (1 - pijNum)) : 0;

				float avgHeight =
						occupancy.at<float>(i, j) > 0 ?
								height.at<float>(i, j)
										/ occupancy.at<float>(i, j) :
								0;
				float pijHeight = 1 - exp(-avgHeight / deltaH);
				float lijHeight =
						pijHeight != 1 ? log(pijHeight / (1 - pijHeight)) : 0;

				float avgProb = wN * lijNum + wH * lijHeight;
				//cout << "occ " << occupancy.at<float>(i,j) << " adjNum " << adjustedNum << " adj/deltan " << adjustedNum/deltaN << " expadjdeltan " << exp(-adjustedNum/deltaN) << " pijnum " << pijNum << " lijnum " << lijNum << " avgheight "
				//		<< avgHeight << " avgheightdeltah " << avgHeight/deltaH << " expavghdeltah " << exp(-avgHeight/deltaH) << " pijheight " << pijHeight << " lijheight " << lijHeight << " avgprob " << avgProb << " deltah " << deltaH << " deltaN " << deltaN  << endl;
				//cout << occupancy.at<float>(i, j) << "|" << pijNum << "|" << lijNum << endl;
				if (avgProb < nt) {
					dispOccupancy.at<cv::Vec3b>(i, j) = cv::Vec3b(0, 0, 0);
				} else if (lijNum >= lt) {
					dispOccupancy.at<cv::Vec3b>(i, j) = cv::Vec3b(0, 0, 255);
				} else {
					dispOccupancy.at<cv::Vec3b>(i, j) = cv::Vec3b(255, 0, 0);
				}
			} else {
				dispOccupancy.at<cv::Vec3b>(i, j) = cv::Vec3b(0, 0, 0);
			}
		}
	}

	cv::Mat mask;
	cv::inRange(dispOccupancy, cv::Scalar(255, 0, 0), cv::Scalar(255, 0, 0),
			mask);
	dispOccupancy.setTo(cv::Scalar(0, 0, 0), mask);

	cv::Mat kernel = cv::getStructuringElement(cv::MORPH_RECT, cv::Size(7, 7));
	cv::morphologyEx(dispOccupancy, dispOccupancy, cv::MORPH_OPEN, kernel);

	double cellWidth = (xRange.end - xRange.start) / occupancy.cols;
	int dilationN = (robotWidth / 2 + clearance) / cellWidth;

	cv::Mat dilationKernel = cv::getStructuringElement(cv::MORPH_RECT,
			cv::Size(2 * dilationN + 1, 1));
	cv::dilate(dispOccupancy, dispOccupancy, dilationKernel);
}

OccupancyGrid::~OccupancyGrid() {
	// TODO Auto-generated destructor stub
}

File: opencv-python/util/OccupancyGrid.py
import numpy as np
import numpy.ma as ma
import scipy as sp
import cv2
from .util_fxn import load_mat
from math import floor, sqrt, exp, log
import time

'''
Translated from C++ occupancy grid implementation by Guilherme Meira

For more information on the algoritm used, look at his Master's Thesis
https://web.wpi.edu/Pubs/ETD/Available/etd-042616-234937/unrestricted/msthesis-guilhermemeira-finalversion.pdf
'''
class OccupancyGrid:
    def __init__(self, vision, occupancy_params):
        self.__vision = vision
        self.__cfg = occupancy_params['cfg']
        occupancy_params = occupancy_params['o_props'][self.__cfg]
        self.__params = occupancy_params
        # Disparity-to-Depth Matrix
        self.q_mat = vision.q
        # Dimensions of the occupancy grid
        self.occupancy_size = tuple(occupancy_params['occupancySize'])
        # 3-D bounds of the space used to form the occupancy grid
        self.x_range = tuple(occupancy_params['xRange'])
        self.y_range = tuple(occupancy_params['yRange'])
        self.z_range = tuple(occupancy_params['zRange'])
        # The y position of the camera
        self.cam_h = occupancy_params['cameraHeight']
        self.r = occupancy_params['r']
        self.c = occupancy_params['c']
        # Probability coefficients
        self.delta_n = occupancy_params['deltaN']
        self.delta_h = occupancy_params['deltaH']
        # Weights
        self.w_n = occupancy_params['wN']
        self.w_h = occupancy_params['wH']
        # Limits for occ status
        self.nt = occupancy_params['nt']
        self.lt = occupancy_params['lt']
        # Robot info
        self.robot_width = occupancy_params['robotWidth']
        self.clearance = occupancy_params['clearance']

    def update(self):
        disparity = self.__vision.disparity
        # Create point cloud
        image3d = cv2.reprojectImageTo3D( \
            disparity, self.q_mat, handleMissingValues=True)
        occupancy = np.zeros(self.occupancy_size)
        height = np.zeros(self.occupancy_size)
        # Final Occupancy Grid
        disp_occ = np.zeros(occupancy.shape, np.int16)

        start = time.clock()
        # pt.y is the y offset of the point relative the the camera
        # it is also the opposite of the actual value (down is +)
        # Scale Z+X to [0,1]

        s_pts = np.dstack((
            (image3d[:,:,0] - self.x_range[0]) / (self.x_range[1] - self.x_range[0]),
            self.cam_h - image3d[:,:,1],
            (image3d[:,:,2] - self.z_range[0]) / (self.z_range[1] - self.z_range[0])
        ))

        invalid = np.any(np.dstack((
            s_pts[:,:,0] < 0, s_pts[:,:,0] > 1,
            s_pts[:,:,2] < 0, s_pts[:,:,2] > 1,
            s_pts[:,:,1] < self.y_range[0], s_pts[:,:,1] > self.y_range[1]
        )), axis=2)

        scaledCoords = np.dstack((
            self.occupancy_size[0] - np.floor(s_pts[:,:,2] * self.occupancy_size[0]),
            np.floor(s_pts[:,:,0] * self.occupancy_size[1])
        )).astype(int)
        span = time.clock() - start
        print('Mat Time = {0}'.format(span))

        start = time.clock()
        for pt, h in zip(scaledCoords[~invalid], s_pts[:,:,1][~invalid]):
            height[pt[0],pt[1]] += h
            occupancy[pt[0],pt[1]] += 1
        span = time.clock() - start
        print('Loop1 Time = {0}'.format(span))

        # Location of camera on occupancy grid
        x_cam = self.occupancy_size[0] / 2
        y_cam = self.occupancy_size[1]

        start = time.clock()
        # For each cell in grid
        for i in range(occupancy.shape[0]):
            for j in range(occupancy.shape[1]):
                # Is there stuff at this cell?
                if occupancy[i,j] > 0:
                    # Get local copy of coords
                    x_pt = j
                    y_pt = i
                    # Euler's formula
                    dist_to_cam = sqrt((x_cam - x_pt) ** 2 +
                        (y_cam - y_pt) ** 2)
                    # Adjust the number of points in cell using sigmoid fxn
                    # r and c are control coefficients
                    adjusted_num = occupancy[i,j] * self.r / (1 + exp(-dist_to_cam * self.c))
                    # Calculate probability of cell being occupied,
                    # based on adjusted number of cell occupants

                    # delta N is a parameter for the probability calculation
                    pij_num = 1 - exp(-(adjusted_num / self.delta_n))
                    # Convert probability to 'log-odds' (logit)
                    lij_num = log(pij_num / (1 - pij_num)) if pij_num != 1 else 0

                    # Get average height in cell
                    avg_height = height[i,j] / occupancy[i,j] if occupancy[i,j] > 0 else 0
                    # Calculate probability of cell being occupied,
                    # based on average height of cell occupants

                    # delta H is a parameter for the probability calculation
                    pij_height = 1 - exp(-avg_height / self.delta_h)
                    # Convert probability to 'log-odds' (logit)
                    lij_height = log(pij_height / (1 - pij_height)) if pij_height != 1 else 0

                    # Estimate the probability that something is in the cell
                    # This is a weighted average, so w_n + w_h = 1
                    avg_prob = self.w_n * lij_num + self.w_h * lij_height

                    if avg_prob < self.nt: # Nothing here (probably)
                        disp_occ[i,j] = 0
                    elif lij_num >= self.lt: # Cell is occupied (probably)
                        disp_occ[i,j] = 2
                    else: # We don't know exactly
                        disp_occ[i,j] = 1
                else: # There is definitely nothing here
                    disp_occ[i,j] = 0
        span = time.clock() - start
        print('Loop2 Time = {0}'.format(span))

        # We don't care about the 'unknowns'
        disp_occ[disp_occ == 1] = 0

        # Get rid of small clusters by eroding, then dilating
        kernel = cv2.getStructuringElement(cv2.MORPH_RECT, (5, 5))
        disp_occ = cv2.morphologyEx(disp_occ, cv2.MORPH_OPEN, kernel)

        # Calculate width of cell relative to original 3d box width
        cell_width = (self.x_range[1] - self.x_range[0]) / occupancy.shape[1]
        # Calculate dilation factor ('radius' of robot + bit extra)
        # Then convert from width in terms of box to width in terms of cells
        dilation_n = int((self.robot_width / 2 + self.clearance) / cell_width)

        # Dilate to ensure clearance
        dilation_kernel = cv2.getStructuringElement(cv2.MORPH_RECT, (2 * dilation_n + 1, 1))
        #disp_occ = cv2.dilate(disp_occ, dilation_kernel)

        # Return the occupancy grid and the point cloud
        self.occupancy = disp_occ
